### PR TITLE
Source layout and implicit namespace package

### PR DESCRIPTION
**Pull request by abn**
This change moves the code base to using src layout. Additionally, this change also converts the `poetry` package to be an implicit namespace package.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -3,7 +3,7 @@ repos:
     rev: 21.9b0
     hooks:
       - id: black
-        exclude: ^poetry/core/_vendor
+        exclude: ^src/poetry/core/_vendor
```
Comment (branchvincent) [inline .pre-commit-config.yaml:6]: unrelated to this pr, but `^src/poetry/core/_vendor` could be moved to the [top-level `exclude`](https://pre-commit.com/#pre-commit-configyaml---top-level)

Comment (abn) [reply .pre-commit-config.yaml:6]: ack, we can do that in another PR. I think @neersighted is doing a general cleanup once the config for `poetry` stabilises.

Comment (neersighted) [reply .pre-commit-config.yaml:6]: I'm personally fine leaving it for now until we start bringing poetry's `.pre-commit-config.yaml` over.

File: pyproject.toml [hunk 2]
```diff
@@ -19,7 +19,7 @@ classifiers = [
 ]
 
 packages = [
-    {include = "poetry"},
+    {include = "poetry", from = "src"},
```
Comment (neersighted) [inline pyproject.toml:22]: Let's add whitespace inside the curl braces here.

**Review by neersighted — CHANGES_REQUESTED** [review]

**Comment by neersighted** [comment]
I'd also squash all the comments down, and then this LGTM.

**Comment by abn** [comment]
> I'd also squash all the comments down, and then this LGTM.

I squashed the tox change in because I guess it is kinda related to the src layout change :) The namespace package change needs to be separate.

**Review by neersighted — APPROVED** [review]

**Pull request merged**